l2-contracts/contracts/AttesterRegistry.sol

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+contract AttesterRegistry {
+    // Owner of the contract (the Authority contract).
+    address public owner;
+
+    // A map of indexes => attesters (used for efficient lookups).
+    mapping(uint256 => Attester) public attesters;
+    // Array to keep track of attester indexes (used for iterations).
+    uint256[] public attesterIndexes;
+    // The last unique value used as an attester index (used for supporting arbitrary removals).
+    uint256 public attesterSequence;
+
+    // The current committee list. Weight and public key are stored explicitly
+    // since they might change after committee selection.
+    CommitteeAttester[] public committee;
+    // The committee list for the next epoch.
+    CommitteeAttester[] public nextCommittee;
+
+    // The current epoch number.
+    uint256 public epoch;
+    // The number of epochs that an attester must be inactive before being
+    // possible to remove it. Needs to be at least 2, in order to guarantee
+    // that any attester in the current or next committee lists is still
+    // available in the registry.
+    uint256 public constant INACTIVITY_DELAY = 2;
```

Comment:
can you please elaborate the motivation for this `isInactive` flag? What exactly imposing this delay changes in the security model?

Reply:
I'd also love to read the docs on the broader design about how `setCommittee` and epoch works. 

I imagine this delay works like this:

| epoch N-3    | epoch N-2     | epoch N-1      | epoch N      |
|--------------|---------------|----------------|--------------|
|              | committee     | next committee |              |
| remove       |               |                | inactivate   |

So the current `epoch` is N, and which means if you `inactivate` someone they will be `inactivatedSince` N. `epoch` is increased in `setCommittee`, so the `committee` and `nextCommittee` are 2 and 1 epoch behind the current one. We don't want to allow `remove` to get rid of members who might be part of those committees. 

I'm not clear though how the `setCommittee` should interact with the actual consensus. Can the owner call `setCommittee` rapidly a few times, resulting in `committee` being forgotten?

Reply:
If that table above is correct, shouldn't `require(epoch >= attesters[idx].inactiveSince + INACTIVITY_DELAY,` below require `epoch > inactiveSince + DELAY` instead of `>=`? 

If `committee` is a snapshot of `epoch-2`, then if e.g. Alice was inactivated in epoch 10, she can be removed in epoch 12, which means she could still be part of the `committee`? 

I could be misunderstanding it, just wanted to clarify 🙏

Reply:
> Or perhaps nextCommittee is just used to allow the new members into the tightly connected validator network.

Speaking of which, I understood the need for `nextCommittee` in the context of validators, because the validator network is separate from the gossip network, and the next cohort of validators have to be allowed to join before they actually have to start voting. 

The attesters, however, operate only on the gossip network, and they have to be known for L1. What is the role of `nextCommittee` for them?

Also naive question: do these contracts live on L2 or L1? The title says L2, but I wonder if it means that's where it's deployed, and if so, how does L1 know which are the current attesters?

Reply:
@RomanBrodetski We can't remove attesters right in the middle of an epoch, so a delay is necessary. A delay of two epochs is the minimum necessary if we want to do rewards/slashing later.
@aakoshh I think you are misunderstanding. First you mark your attester as inactive, then you remove it.
The contracts live on L2 and the L1 will be aware of the attesters via L2->L1 communication logs. That's outside the scope of this PR though.
NextCommittee might not be strictly necessary for attesters, but it doesn't hurt and this way the contracts are the same for attesters and validators.

Reply:
> I think you are misunderstanding. First you mark your attester as inactive, then you remove it.

I did understand that part 👍 

Sorry if I added to the confusion. I just wanted to visualise that what you mark inactive in epoch N, you should be able to remove in epoch N+3, or vice versa when you are in epoch N, you can remove nodes you inactivated in epoch N-3, and thus not affect the snapshots you took which are in `committee` (epoch N-2) and `nextCommittee` (epoch N-1). So yes, I get that you inactivate and then remove later.

Reply:
Had a re-think and I'm convinced that the `require(epoch >= inactiveSince + 2)` is correct. 

For example:
* epoch=10, Alice.invactive=false, Bob.inactive=false, committee=[Alice, Bob], nextCommittee=[Alice, Bob]
* inactivate(Alice) -> Alice.inactive=true, Alice.inactiveSince=10
* setCommittee() -> epoch=11, committee=[Alice, Bob], nextCommittee=[Bob]
* setCommittee() -> epoch=12, committee=[Bob], nextCommittee=[Bob]
* remove(Alice) -> ok, she's not in any committee